protocols/gossipsub/src/behaviour.rs

```diff
@@ -3126,12 +3199,94 @@ where
     }
 }
 
+/// This is called when peers are added to any mesh. It checks if the peer existed
+/// in any other mesh. If this is the first mesh they have joined, it queues a message to notify
+/// the appropriate connection handler to maintain a connection.
+fn peer_added_to_mesh(
+    peer_id: PeerId,
+    new_topics: Vec<&TopicHash>,
+    mesh: &HashMap<TopicHash, BTreeSet<PeerId>>,
+    known_topics: Option<&BTreeSet<TopicHash>>,
+    events: &mut VecDeque<GossipsubNetworkBehaviourAction>,
+    connections: &HashMap<PeerId, PeerConnections>,
+) {
+    // Ensure there is an active connection
+    let connection_id = {
+        let conn = connections.get(&peer_id).expect("To be connected to peer.");
+        assert!(
+            !conn.connections.is_empty(),
+            "Must have at least one connection"
+        );
+        conn.connections[0]
+    };
+
+    if let Some(topics) = known_topics {
+        for topic in topics {
+            if !new_topics.contains(&topic) {
+                if let Some(mesh_peers) = mesh.get(topic) {
+                    if mesh_peers.contains(&peer_id) {
+                        // the peer is already in a mesh for another topic
+                        return;
+                    }
+                }
+            }
+        }
+    }
+    // This is the first mesh the peer has joined, inform the handler
+    events.push_back(NetworkBehaviourAction::NotifyHandler {
+        peer_id,
+        event: Arc::new(GossipsubHandlerIn::JoinedMesh),
+        handler: NotifyHandler::One(connection_id),
+    });
+}
+
+/// This is called when peers are removed from a mesh. It checks if the peer exists
+/// in any other mesh. If this is the last mesh they have joined, we return true, in order to
+/// notify the handler to no longer maintain a connection.
+fn peer_removed_from_mesh(
+    peer_id: PeerId,
+    old_topic: &TopicHash,
+    mesh: &HashMap<TopicHash, BTreeSet<PeerId>>,
+    known_topics: Option<&BTreeSet<TopicHash>>,
+    events: &mut VecDeque<GossipsubNetworkBehaviourAction>,
+    connections: &HashMap<PeerId, PeerConnections>,
+) {
+    // Ensure there is an active connection
+    let connection_id = {
+        let conn = connections.get(&peer_id).expect("To be connected to peer.");
+        assert!(
+            !conn.connections.is_empty(),
+            "There should be at least one connection to a peer"
+        );
+        conn.connections[0]
+    };
```

Comment:
```suggestion
    let connection_id = connections.get(&peer_id)
        .expect("To be connected to peer.")
        .get(0)
        .expect("There should be at least one connection to a peer");
```

This should be valid as well, right?